internal/pkg/encryption/keys/kms.go

```diff
@@ -0,0 +1,165 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package keys
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/tls"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/siderolabs/go-blockdevice/blockdevice/encryption"
+	"github.com/siderolabs/go-blockdevice/blockdevice/encryption/luks"
+	"github.com/siderolabs/go-blockdevice/blockdevice/encryption/token"
+	"github.com/siderolabs/go-retry/retry"
+	"github.com/siderolabs/kms-client/api/kms"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/status"
+
+	"github.com/siderolabs/talos/internal/pkg/endpoint"
+	"github.com/siderolabs/talos/internal/pkg/smbios"
+)
+
+// KMSToken is the userdata stored in the partition token metadata.
+type KMSToken struct {
+	SealedData []byte `json:"sealedData"`
+}
+
+// KMSKeyHandler seals token using KMS service.
+type KMSKeyHandler struct {
+	key
+	kmsEndpoint string
+}
+
+// NewKMSKeyHandler creates new KMSKeyHandler.
+func NewKMSKeyHandler(key key, kmsEndpoint string) (*KMSKeyHandler, error) {
+	return &KMSKeyHandler{
+		key:         key,
+		kmsEndpoint: kmsEndpoint,
+	}, nil
+}
+
+// NewKey implements Handler interface.
+func (h *KMSKeyHandler) NewKey() (*encryption.Key, token.Token, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	conn, err := h.getConn(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error dialing KMS endpoint %q: %w", h.kmsEndpoint, err)
+	}
+
+	client := kms.NewKMSServiceClient(conn)
+
+	s, err := smbios.GetSMBIOSInfo()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var resp *kms.Response
+
+	key := make([]byte, 32)
+	if _, err = io.ReadFull(rand.Reader, key); err != nil {
+		return nil, nil, err
+	}
+
+	err = retry.Exponential(time.Minute*5, retry.WithUnits(time.Second), retry.WithJitter(time.Second)).RetryWithContext(ctx, func(ctx context.Context) error {
```

Comment:
maybe a minute is enough? :thinking:

Reply:
actually maybe even 30 seconds is enough. updated